### ISSUE TITLE
Incorrect tokenization with tokens added using tokenizer.add_tokens()

### ISSUE THREAD

**Issue by Muks14x** [huggingface/transformers#7549]
## Environment info

     
- `transformers` version: 3.0.2
- Platform: macOS-10.15.6-x86_64-i386-64bit
- Python version: 3.8.3
- PyTorch version (GPU?): 1.5.1 (False)
- Tensorflow version (GPU?): 2.3.0 (False)
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help

It's a tokenization issue, so tagging @mfuntowicz
Also happens with rust tokenizers, so tagging @n1t0 

## Information

Model I am using (Bert, XLNet ...): RoBERTa (but happens anywhere)

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [x] irrelevant to the bug
* [ ] an official GLUE/SQUaD task: (give the name)
* [ ] my own task or dataset: (give details below)

## To reproduce

```python
>>> from transformers import RobertaTokenizer
>>> tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
>>> tokenizer.add_tokens(['\U00030001', '\U00030002', '\U00030002\U00030001'])
3
>>> tokenizer.tokenize('\U00030002\U00030001')
['\U00030002\U00030001']
>>> tokenizer.tokenize('\U00030001\U00030002\U00030001') ## produces incorrect output. the last two tokens should've been together and should not have gotten split
['\U00030001', '\U00030002', '\U00030001']
>>> tokenizer.unique_no_split_tokens
['\U00030001', '<s>', '</s>', '<unk>', '\U00030002\U00030001', '<mask>', '<pad>', '\U00030002']
>>> tokenizer.unique_no_split_tokens.sort(key=lambda x: -len(x)) ## On sorting the unique_no_split_tokens by the lengths, this seems to get fixed. I suspect that internally the code is checking the presence of added tokens in this order?
>>> tokenizer.unique_no_split_tokens
['<mask>', '<unk>', '<pad>', '</s>', '<s>', '\U00030002\U00030001', '\U00030001', '\U00030002']
>>> tokenizer.tokenize('\U00030001\U00030002\U00030001')
['\U00030001', '\U00030002\U00030001']
```

## Expected behavior

Tokenization seems to depend on the order in which tokens get added to the model. Just to show what happens, I've added some (very high valued) unicode character tokens and run the tokenization. 

Basically, '\U00030002', '\U00030001' got split the first time, which should not have happened since ''\U00030002\U00030001' is part of the vocabulary. On sorting the tokenizer.unique_no_split_tokens list by length, it seems to fix this issue. This makes me uneasy using add_tokens now with tokens that share overlaps.

Also, the problem persists with the rust tokenizers library (RobertaTokenizerFast). But I don't want to open up another issue without first making sure that this is an issue.

**Comment by LysandreJik** [comment]
Pinging @n1t0 for advice.
